design/private-repos.md

```diff
@@ -0,0 +1,72 @@
+# Screwdriver Access to Private Repositories
+
+## Context
+Provide a way for users to create and execute a pipeline for a private GitHub repository.
+
+This documentation covers the design proposal for that new feature.
+
+## Status
+
+- Oct 10, 2018: Design document
+
+## The Problem
+
+The current authentication methods do not provide access to private repositories.  This includes
+GitHub API calls as well as git clones via SSH.
+
+The OAuth token provided today for GitHub API calls is limited in scope to only public repositories.  If this scope were to 
+be expanded, then all users (whether they need this expanded scope or not) would be granting Screwdriver
+a security scope larger than required.  The screwdriver app should only be granted scope
+that is required.
+There are several stages of a Screwdriver pipeline in which this expanded scope is required
+when dealing with private repositories.  This document will outline each stage and how it will be addressed.
+
+Additionally, the Screwdriver app only has access to clone public repositories with the ssh key
+currently being used by Screwdriver.
+
+## Background
+
+Reference [Issue 1079][issue1079]
+
+## Proposal
+Any user intending to use Screwdriver with private repositories must perform the following prerequisites:
+- create a Personal Access Token within the Developer Settings of their Github user profile
+and include that token into their Screwdriver user profile.
+	- Github Personal Access Token is created at: https://github.com/settings/tokens/new
+	- Screwdriver user profile is edited at:  /user-settings
+- create/obtain an ssh keypair 
+	- store the public key within the Deploy Keys section of the private repository settings 
+	- associate the private key with the Screwdriver pipeline associated with this private repository
+
+Private Repository Pipeline Creation:
+1 A GitHub Personal Access Token that allows control of private repositories must be associated to the user profile within Screwdriver.
```

Comment:
One concern here is that this Personal Access Token has wide permission and can do anything including delete user stuff. We are back with original concern of preventing SD having access to a user token which can do pretty much anything on behalf of user.